---
title: "Ranking with TPM vs Count inputs with or without shrinkage" 
output: html_document
author: "Mira Sohn"
---

## Raw data 

#### - **Link**: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157852

#### - **Study aim**: Reveal neurological complications in patients with COVID-19

#### - **Study summary**: Human Pluripotent Stem Cell-Derived Neural Cells and Brain Organoids were infected with SARS-CoV-2 for 72hrs and analyzed with bulk RNA-seq.

#### - Samples: Mock- or SARS-CoV-2-infected neuronal cells 

```{r global_options, include=FALSE}
knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE
    )
```



## Loading packages

```{r loading_packages}
library(data.table)
library(rmarkdown)
library(AnnotationHub)
library(tidyverse)
library(tximport)
library(ggplot2)
library(DESeq2)
library(UpSetR)
```

## Setting AnnotationHub


```{r annotationhub_setup}
AnnotationSpecies <- "Homo sapiens"  # Assign your species 
ah <- AnnotationHub(hub=getAnnotationHubOption("URL"))   # Bring annotation DB
```

## Running AnnotationHub 

```{r run_annotationhub}

ahQuery <- query(ah, c("OrgDb", AnnotationSpecies))      # Filter annotation of interest

if (length(ahQuery) == 1) {
    DBName <- names(ahQuery)
} else if (length(ahQuery) > 1) {
               DBName <- names(ahQuery)[1]
} else {
    print("You don't have a valid DB")
    rmarkdown::render() 
} 

AnnoDb <- ah[[DBName]] # Store into an OrgDb object  

# Explore your OrgDb object with following accessors:
# columns(AnnpDb)
# keytypes(AnnoDb)
# keys(AnnoDb, keytype=..)
# select(AnnoDb, keys=.., columns=.., keytype=...)
AnnoKey <- keys(AnnoDb, keytype="ENSEMBLTRANS")

# Note: Annotation has to be done with not genome but transcripts 
AnnoDb <- select(AnnoDb, 
                 AnnoKey,
                 keytype="ENSEMBLTRANS",
                 columns="ENSEMBL")

# Check if your AnnoDb has been extracted and saved correctely
class(AnnoDb)
head(AnnoDb)
```




## Defining file name and path for .sf files

#### .sf files have been created from fastq data by salmon


```{r preparing_importing.sf}

# This code chunk needs to be revised by yourself 

# Define sample names 
SampleNames <-  c("Mock_72hpi_S1",
                 "Mock_72hpi_S2",
                 "Mock_72hpi_S3",
                 "SARS-CoV-2_72hpi_S7",
                 "SARS-CoV-2_72hpi_S8",
                 "SARS-CoV-2_72hpi_S9") 

# Define group level
GroupLevel <- c("Mock", "COVID")

# Define contrast for DE analysis
Contrast <- c("Group", "COVID", "Mock")

# Define a vector for comparing TPM vs Counts effect 
TvC <- c("TPM", "Counts")

# Define .sf file path
sf <- c(paste0(SampleNames,
               ".salmon_quant/quant.sf"))

# Define sample groups
group <- c(rep("Mock", 3), rep("COVID", 3))

# Create metadata
metadata <- data.frame(Sample=factor(SampleNames, levels=SampleNames),
                       Group=factor(group, levels=GroupLevel),
                       Path=sf)

rownames(metadata) <- SampleNames

# Explore the metadata
print(metadata)
```

## Converting .sf files to txi list 

#### - txi_tpm: stores **TPM** with the argument **"countsFromAbundance="lengthScaledTPM"**

#### - txi_counts: stores **original counts** 

#### - Despite that the TPM matrix is not recommended as an input for DE analysis due to the fact that it doesn't account gene length, TPM stored in a tximport (by **tximport(..., countsFromAbundance="lengthScaledTPM")**) can be used for DE analysis by being inputted with the **DESeqDataSetFromTximport()** funtion in DESeq2 workflow as the gene length is automatically adjusted by DESeqDataSetFromTximport(). 

#### - In this project, two txi objects were created with or without the countsFromAbundance="lengthScaledTPM" argument and compared in downstream DE analysis. 

#### - If you don't want gene-level summarization, set **txOut=TRUE**. 

#### - References: [**tximport doc**](https://bioconductor.riken.jp/packages/3.4/bioc/vignettes/tximport/inst/doc/tximport.html), [**DESeq2 doc "Why unnormalized counts?"**](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts), [**Soneson et al. 2016**](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4712774/), [**Developer Dr. Love's comment**](https://support.bioconductor.org/p/84059/), [**Harvard Chan Bioinformatics Core workshop**](https://github.com/hbctraining/DGE_workshop_salmon_online/blob/master/lessons/02_DGE_count_normalization.md)


```{r saving_reads_to_dataframe}

# Assign sample names to the input (.sf) file path
names(sf) <- SampleNames

# Run tximport
# tpm vs original counts
# input sf: a factor of all .sf files' path
txi_tpm <- tximport(sf, 
                    type="salmon",
                    tx2gene=AnnoDb,
                    countsFromAbundance="lengthScaledTPM", # Extracts TPM 
                    ignoreTxVersion=T) 

txi_counts <- tximport(sf, 
                    type="salmon",
                    tx2gene=AnnoDb,
                    ignoreTxVersion=T) 
```

## Exploring the txi outputs 

```{r txi_outputs}
# tpm 
head(txi_tpm$counts)
dim(txi_tpm$counts)

# counts
head(txi_counts$counts)
dim(txi_counts$counts)
```


## Creating DESeq objects from txi and VST

#### - Note: The tximport-to-DESeq2 approach uses estimated gene counts from the transcript abundance quantifiers, but not normalized counts.

#### - The **DESeqDataSetFromTximport()** function generated an DESeq object (aka **dds**) with the **txi input**. 

#### - References: [**DESeq2 doc "Transcript abundance files"**](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#tximport), [**DESeq2 doc "Variance stabilizing transformation"**](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#variance-stabilizing-transformation)




```{r creating_dds}

# Initialize a list storing dds object
desList <- list()

for (x in TvC) {

    if (x == TvC[1]) {

        desList[[x]] <- DESeqDataSetFromTximport(txi_tpm, 
                                    colData=metadata,
                                    design=~Group)
    } else {

        desList[[x]] <- DESeqDataSetFromTximport(txi_counts, 
                                    colData=metadata,
                                    design=~Group)

    }
}

# Exploring created dds 
# TPM 
desList[[TvC[1]]]
head(counts(desList[[TvC[1]]]))

# Counts
desList[[TvC[2]]]
head(counts(desList[[TvC[2]]]))

```




## Estimating size factors, dispersions, and conducting Wald Test

#### - **Dispersion** is calculated as a **measure of variation** instead of variance since variance gets larger when gene expression gets higher. 

#### - **Wald test** is the default setting of DESeq2 which tests null hypothesis between **two groups**. You should use **Likelihood ratio test (LRT)** when comparing **more than two groups**.     

#### - Messages when "Counts <- DESeqPrep_fn(Counts)" was run:       
using **'avgTxLength'** from assays(dds), correcting for library size
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates

#### - References: [**Harvard Chan Bioinformatics Core workshop I**](https://github.com/hbctraining/DGE_workshop_salmon_online/blob/master/lessons/05b_wald_test_results.md), [**Harvard Chan Bioinformatics Core workshop II**](https://github.com/hbctraining/DGE_workshop_salmon_online/blob/master/lessons/05a_hypothesis_testing.md), [**Harvard Chan Bioinformatics Core workshop III**](https://github.com/hbctraining/DGE_workshop_salmon_online/blob/master/lessons/04b_DGE_DESeq2_analysis.md), [**DESeq2 "Wald test indivisual steps**](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#wald-test-individual-steps), [**DESeq2 doc "Likelihood ratio test"**](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#likelihood-ratio-test)

```{r DESeq_prep}



# Update dds 

for (x in TvC) { 

    desList[[x]] <- estimateSizeFactors(desList[[x]])
    desList[[x]] <- estimateDispersions(desList[[x]])
    desList[[x]] <- nbinomWaldTest(desList[[x]])
}

# Explore the updates 
desList
```


## Running DE analysis without shrinkage 

```{r DE_analysis}

# Create a list for TPM and Counts dds 
ddsList <- desList  # DE without shrinkage  

for (x in TvC) {
    
    # Run DESeq() 
    ddsList[[x]] <- DESeq(desList[[x]])
    print(resultsNames(ddsList[[x]]))


}


```


## Extracting DE results with or without shrinkage

#### - The **alpha** denotes threshold of **false discovery rate (FDR)** assigned by users.

#### - In this analysis, the alpha is set to **0.1** 


```{r extracting_results}

# Set FDR threshold 
alpha=0.1 

# FDR threshold vector
FDRv=c(paste("<",  alpha), 
       paste(">",  alpha)) 

# Initialize lists of result tables 
resList <- ddsList 

# Set a function cleaning table
Sig.fn <- function(df, Input) {
    
    df <- df %>% 
        rownames_to_column(var="Gene") %>%
        mutate(FDR=ifelse(padj < 0.1 & !is.na(padj), 
                                   FDRv[1], 
                                   FDRv[2]), 
               Input=Input) 
    return(df)
}


for (x in TvC) {

        # Extract data frames from unshrunken lfc & clean data 
        resList[[x]] <- as.data.frame(results(ddsList[[x]], 
                                              contrast=Contrast, 
                                              alpha=alpha)) %>% Sig.fn(x)

         } 





# Explore the results 
summary(resList)
head(resList[[TvC[1]]])
head(resList[[TvC[2]]])

```




## Exploring mean-difference relationship with MA plots


#### - **x-axis**: expression level (baseMean))

#### - **y-axis**: fold change (log2FoldChange)

#### - **Red dashed lines**: log2FoldChange = -1 and 1


#### - Reference: [**DESeq2 doc "MA-plot"**](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#ma-plot)


```{r MAplot}

# Set min log2 fold change of interest 
mLog <- c(-1, 1)


# Create an input data frame for MA plots 
both.df <- rbind(resList[[TvC[1]]], resList[[2]])

# Create MA plots
ggplot(both.df, 
       aes(x=baseMean, y=log2FoldChange, color=FDR)) +
geom_point() + scale_x_log10() + facet_grid(~Input) + 
theme_bw() + scale_color_manual(values=c("blue", "grey")) + 
ggtitle(paste("MA plot by Input")) + geom_hline(yintercept=c(mLog[1], mLog[2]), linetype="dashed", color="red") 


```


## Exploring distribution of log2FoldChange by input type

#### - Black dashed lines: log2FoldChange = -1 and 1 

#### - **x-axis**: gene expression level (log2FoldChange) 

#### - **y-axis**: number of genes 

```{r L2FC_distribution}


# Create lfc distribution plots
ggplot(both.df[both.df$FDR == FDRv[1],],                                 
       aes(x=log2FoldChange, color=Input)) + 
geom_density(size=1, aes(y=..count..)) + theme_bw() + ylab("Count") + geom_vline(xintercept=c(mLog[1], mLog[2]), color="black", linetype="dashed", size=1) + ggtitle("Distribution of Log2 Folds by Input Type") +
xlim(-10, 10)

```

## NA statistics: zero count genes & outlier genes

#### When NAs appear in  

#### - **log2FoldChange**: zero counts in all samples

#### - **padj**: too little information 

#### - **pval & padj**: at least one replicate was an outlier 

```{r NA_genes}

# Count number of NA genes  
type=c("Zero Counts", "Outliers", "Total NA Genes") 

# Create a data frame storing number of NA genes by type
NA.genes <- both.df %>% 
    group_by(Input) %>% 
    summarize(zero=sum(is.na(log2FoldChange)), 
              outlier=sum(is.na(pvalue) & is.na(padj))) %>% 
    mutate(total=zero + outlier) %>%
    gather(Type, Number, -Input) %>% 
    mutate(Type=factor(case_when(Type == "zero" ~ type[1],
                                 Type == "outlier" ~ type[2],
                                 Type == "total" ~ type[3]),
                       levels=type))

# Plot number of NA genes 
ggplot(NA.genes, 
       aes(x=Type, y=Number, group=Input, fill=Input, label=Number)) + 
    geom_bar(stat="identity", position="dodge") + 
    theme_bw() +
    geom_text(position=position_dodge(width=1), vjust=1.5) + 
    ggtitle("Number of NA Genes") + 
    ylab("Number of Genes")


```

## Ranking DEGs 

#### - **fdr.rank**: ranked by FDR

#### - **lfc.rank**: ranked by absolute fold change 

#### - **up.lfc.rank**: ranked by magnitude of fold change increase 

#### - **down.lfc.rank**: ranked by manitude of fold change decrease


```{r determining_DEGrank}

# Create a new list having DE table with FDR below alpha
fdr.rank <- resList
lfc.rank <- resList
up.lfc.rank <- resList
down.lfc.rank <- resList

# Set a function cleaning a data frame 
filter.fdr.fn <- function(df) {as.data.table(df[df$FDR == FDRv[1],])}

# Set a function creating a column for the rank
Ranking.fn <- function(x) {mutate(x, Rank=1:nrow(x))}



for (x in TvC) {

    rdf <- resList[[x]]

    fdr.rank[[x]] <- filter.fdr.fn(rdf) %>% arrange(padj) %>% Ranking.fn()

    lfc.rank[[x]] <- filter.fdr.fn(rdf) %>% arrange(desc(abs(log2FoldChange))) %>% Ranking.fn()

    up.lfc.rank[[x]] <- filter.fdr.fn(rdf) %>% arrange(desc(log2FoldChange)) %>% Ranking.fn() 

    down.lfc.rank[[x]] <- filter.fdr.fn(rdf) %>% arrange(log2FoldChange) %>% Ranking.fn()

}

# Explore the ranking outputs
head(fdr.rank[[1]])
head(lfc.rank[[1]])
head(up.lfc.rank[[1]])
head(down.lfc.rank[[1]])

```

## Calculating rank difference 


```{r determining_rankdiff}




# Set a function rebuilding DE tables with gene ranks 
rankdiff.fn <- function(List){

    # Select columns and join the data frames by gene
    full_join(List[[TvC[1]]][, .(Gene, Input, Rank, baseMean)], 
              List[[TvC[2]]][, .(Gene, Input, Rank, baseMean)], 
              by="Gene") %>%
    
    # Add columns assining gene expression levels, rank differences, and mean ranks
    mutate(logMeanExpression=log(baseMean.x+baseMean.y/2),
           RankDiff=Rank.x-Rank.y, 
           MeanRank=(Rank.x+Rank.y)/2)
} 

# Calculate rank difference by ranking type
fdr.rankdiff <- rankdiff.fn(fdr.rank)
lfc.rankdiff <- rankdiff.fn(lfc.rank)
up.lfc.rankdiff <- rankdiff.fn(up.lfc.rank)
down.lfc.rankdiff <- rankdiff.fn(down.lfc.rank)

# Explore the calculated rank differences
head(fdr.rankdiff)
head(lfc.rankdiff)
head(up.lfc.rankdiff)
head(down.lfc.rankdiff)




```


## Visualizing DEG ranks I: TPM- vs Counts-input

#### - **x-axis**: rank with TPM input

#### - **y-axis**: rank with Counts input 

#### - Black diagonal lines: rank with TPM = rank with Counts 

#### - Dot color: gene expression level (log-baseMean)

#### - 16 genes were missing in the plots

```{r ranking_plot}

# Set a function plotting gene ranks between TPM- (x-axis) and Counts-Inputs (y-axis)
ranking.plot.fn <- function(df, rankedby) {


    ggplot(df, 
           aes(x=Rank.x, y=Rank.y, color=logMeanExpression)) + geom_point(alpha=0.5) + theme_bw() + theme(strip.text.x=element_text(size=10)) + xlab("Rank with TPM") + ylab("Rank with Counts") + geom_abline(slope=1, color="black", size=0.5) + ggtitle(paste(rankedby, "Ranking with TPM vs Count Inputs")) + scale_color_gradient(low="blue", high="red") 
}

# Print output plots
ranking.plot.fn(fdr.rankdiff, "FDR")
ranking.plot.fn(lfc.rankdiff, "Log2FoldChange")
ranking.plot.fn(up.lfc.rankdiff, "Log2FoldChange (Increase)")
ranking.plot.fn(down.lfc.rankdiff, "Log2FoldChange (Decrease)")

```



## Visualizing DEG ranks II: Rank difference 

#### - **x-axis**: expression level (log-baseMean)

#### - **y-axis**: rank difference (rank with TPM - rank with Counts)

#### - **black horizontal lines**: rank with TPM = rank with Counts 

#### - **dot color**: average rank 

#### - 16 genes were missing in the plots

```{r co`mparing_DEGrankII}

# Set a function plotting the rank difference over the gene expression level
rankdiff.plot.fn <- function(df, rankedby) {


    ggplot(df, aes(x=logMeanExpression, y=RankDiff, color=MeanRank)) + 
        geom_point(alpha=0.5) + 
        theme_bw() + 
        theme(strip.text.x=element_text(size=10)) + 
        ylab("Rank Difference (TPM - Count)") + 
        ggtitle(paste("Rank Difference Inputs (TPM - Count):", rankedby)) + 
        geom_hline(yintercept=0, color="black", size=0.5) + scale_color_gradient(low="blue", high="red") +
        ylim(-110, 110)
}

# Print output plots
rankdiff.plot.fn(fdr.rankdiff, "FDR")
rankdiff.plot.fn(lfc.rankdiff, "Log2FoldChange")
rankdiff.plot.fn(up.lfc.rankdiff, "Log2FoldChange (Increase)")
rankdiff.plot.fn(down.lfc.rankdiff, "Log2FoldChange (Decrease)")



```

## Distribution of rank difference  

#### - **y-axis**: abs(TPM-Count inputs) 

#### - **x-axis**: FDR or log2FoldChange (Increase/Decrease)

#### - **red dashed line**: Rank difference = 25

#### - **blue dashed line**: Rank difference = 10


```{r rankdiff_distribution}

# Create a new data frame storing rank difference by result type
rankdiff.dist <- data.frame(FDR=abs(fdr.rankdiff$RankDiff), 
                            log2FoldChange=abs(lfc.rankdiff$RankDiff), 
                            log2FoldChange.Increase=abs(up.lfc.rankdiff$RankDiff),
                            log2FoldChange.Decrease=abs(down.lfc.rankdiff$RankDiff)) %>% gather(Result, RankDiff) 

# Assine result column as a factor with levels 
rankdiff.levels <- c("FDR", 
                     "log2FoldChange", 
                     "log2FoldChange.Increase", 
                     "log2FoldChange.Decrease")

rankdiff.dist$Result <- factor(rankdiff.dist$Result, levels=rankdiff.levels)

# Plot distribution of absolute rank difference
ggplot(rankdiff.dist,
       aes(x=Result, y=RankDiff, color=Result)) +
geom_jitter(alpha=0.5) + 
geom_boxplot(alpha=0.5, fill="grey", color="black") + 
theme_bw() + 
theme(axis.text.x=element_text(angle=45, hjust=1)) +
ggtitle("Distribution of Absolute Rank Difference") + 
ylab("Absolute Rank Difference") + 
geom_hline(yintercept=c(10, 25),     # depends on your interest
           linetype="dashed", 
           color=c("blue", "red"))


```


## Relationship between rank difference and number of transcript versions 

#### - **y-axis**: abs(TPM-Count inputs) 

#### - **x-axis**: number of transcripts (number of transcript id / gene id) 

#### - **dot color**: mean rank

#### - 16 genes were missing in the plots

```{r rankdiff_vs_transcripts}

# Create a data frame storing the number of transcripts by gene id
AnnoDb.ntrans <- AnnoDb %>% 
    group_by(ENSEMBL) %>% 
    summarize(num.trans=n())

# Set a function adding the number of transcripts by gene id 
add.ntrans.fn <- function(df) {

    left_join(df, AnnoDb.ntrans, by=c("Gene"="ENSEMBL"))}

# Create a list storing rankdiff data frames 
rankList <- list(fdr.rankdiff, 
                 lfc.rankdiff, 
                 up.lfc.rankdiff, 
                 down.lfc.rankdiff)

# Name the list 
names(rankList) <- rankdiff.levels

# Add a column indicating the number of transcripts by gene id to every rankdiff data frame
for (x in rankdiff.levels) {

    rankList[[x]] <- add.ntrans.fn(rankList[[x]])
}


# Explore the edited data frames
summary(rankList)
head(rankList[[1]])
head(rankList[[2]])
head(rankList[[3]])
head(rankList[[4]])

# Set a function plotting rank difference vs number of transcripts 
rank.ntrans.plot.fn <- function(df, title) {

    ggplot(df, aes(x=num.trans, y=abs(RankDiff), color=MeanRank)) + 
        geom_jitter(alpha=0.5) + 
        theme_bw() + 
        ggtitle(paste("Rank Difference vs Number of Transcripts in", title)) + 
        xlab("Number of Transcripts") + 
        ylab("Absolute Rank Difference (TPM - Counts)") + scale_color_gradient(low="blue", high="red") 
}

# Print the plots
rank.ntrans.plot.fn(rankList[[rankdiff.levels[1]]], "FDR")
rank.ntrans.plot.fn(rankList[[rankdiff.levels[2]]], "log2FoldChange")
rank.ntrans.plot.fn(rankList[[rankdiff.levels[3]]], "log2FoldChange (Increase)")
rank.ntrans.plot.fn(rankList[[rankdiff.levels[4]]], "log2FoldChange (Decrease)")

```


## Finding genes having large difference in rankings 

```{r large_rankdiff_genes}

# Initialize a list storing rankdiff genes 
large.rankdiff <- rankList

# Assign a vector storing minimum (thresholds) rankdiff for filtering large rankdiff genes
rankdiff.thr <- c(25, 10, 10, 10)

names(rankdiff.thr) <- rankdiff.levels

for (x in rankdiff.levels) {

    # Filter out observations below the rankdiff thresholds
    large.rankdiff[[x]] <- subset(rankList[[x]], 
                                      abs(RankDiff) > rankdiff.thr[x]) 

}

# Explore the filtered genes 
summary(large.rankdiff)

dim(large.rankdiff[[rankdiff.levels[1]]])
dim(large.rankdiff[[rankdiff.levels[2]]])
dim(large.rankdiff[[rankdiff.levels[3]]])
dim(large.rankdiff[[rankdiff.levels[4]]])

head(large.rankdiff[[rankdiff.levels[1]]])
head(large.rankdiff[[rankdiff.levels[2]]])
head(large.rankdiff[[rankdiff.levels[3]]])
head(large.rankdiff[[rankdiff.levels[4]]])

```


## Summarizing up/down DEGs with an upset plot 

#### - **red bar**: input type

#### - **blue bar**: directionality of gene expression change

```{r upsetplotting}

# Set a function cleaning data to generate upset plots 
upset.input.fn <- function(df) {

    df <- df %>% 

        # Filter genes with valid padj 
        filter(!is.na(padj)) %>% 
        
        mutate(Up=ifelse(FDR == FDRv[1] & log2FoldChange > 0, Gene, ""), # What are upregulated genes? 
               
               Down=ifelse(FDR == FDRv[1] & log2FoldChange < 0, Gene, ""),  # What are downregulated genes? 
               
               Unchanged=ifelse(FDR == FDRv[2], Gene, ""),   # What are unchanged genes? 
               
               TPM_Input=ifelse(Input == "TPM", Gene, ""),   # What are the genes from TPM input? 
               
               Counts_Input=ifelse(Input == "Counts", Gene, ""))   # What are the genes from Counts input? 

    # Create a list storing groups of interest
    upsetInput <- list(Up=df$Up, 
                       Down=df$Down, 
                       Unchanged=df$Unchanged, 
                       TPM_Input=df$TPM, 
                       Counts_Input=df$Counts) 

    return(upsetInput)

}

upsetList <- upset.input.fn(rbind(both.df))


# Create the upset plot 
upset(fromList(upsetList), 
      sets=names(upsetList),   # What group to display 
      sets.x.label="Number of Genes per Group",
      order.by="freq",
      point.size=3,
      sets.bar.color=c("red", "red","blue", "blue", "blue"),
      text.scale = 1.5, number.angles=30) 
```





## Session Info 

```{r sessionInfo}
sessionInfo()
```
